packages/xrpl/src/models/transactions/oracleSet.ts

```diff
@@ -0,0 +1,157 @@
+ *
+ * @category Transaction Models
+ */
+export interface OracleSet extends BaseTransaction {
+  TransactionType: 'OracleSet'
+
+  /**
+   * A unique identifier of the price oracle for the Account.
+   */
+  OracleDocumentID: number
+
+  /**
+   * The time the data was last updated, represented as a unix timestamp in seconds.
+   */
+  LastUpdateTime: number
+
+  /**
+   * An array of up to 10 PriceData objects, each representing the price information
+   * for a token pair. More than five PriceData objects require two owner reserves.
+   */
+  PriceDataSeries: PriceData[]
+
+  /**
+   * An arbitrary value that identifies an oracle provider, such as Chainlink, Band,
+   * or DIA. This field is a string, up to 256 ASCII hex encoded characters (0x20-0x7E).
+   * This field is required when creating a new Oracle ledger entry, but is optional for updates.
+   */
+  Provider?: string
+
+  /**
+   * An optional Universal Resource Identifier to reference price data off-chain. This field is limited to 256 bytes.
+   */
+  URI?: string
+
+  /**
+   * Describes the type of asset, such as "currency", "commodity", or "index". This field is a string, up to 16 ASCII
+   * hex encoded characters (0x20-0x7E). This field is required when creating a new Oracle ledger entry, but is optional
+   * for updates.
+   */
+  AssetClass?: string
+}
+
+/**
+ * Verify the form and type of a OracleSet at runtime.
+ *
+ * @param tx - A OracleSet Transaction.
+ * @throws When the OracleSet is malformed.
+ */
+// eslint-disable-next-line max-lines-per-function -- necessary to validate many fields
+export function validateOracleSet(tx: Record<string, unknown>): void {
+  validateBaseTransaction(tx)
+
+  validateRequiredField(tx, 'OracleDocumentID', isNumber)
+
+  validateRequiredField(tx, 'LastUpdateTime', isNumber)
+
+  validateOptionalField(tx, 'Provider', isString)
+
+  validateOptionalField(tx, 'URI', isString)
+
+  validateOptionalField(tx, 'AssetClass', isString)
+
+  // eslint-disable-next-line max-lines-per-function -- necessary to validate many fields
+  validateRequiredField(tx, 'PriceDataSeries', (value) => {
+    if (!Array.isArray(value)) {
+      throw new ValidationError('OracleSet: PriceDataSeries must be an array')
+    }
+
+    if (value.length > PRICE_DATA_SERIES_MAX_LENGTH) {
+      throw new ValidationError(
+        `OracleSet: PriceDataSeries must have at most ${PRICE_DATA_SERIES_MAX_LENGTH} PriceData objects`,
+      )
+    }
+
+    // TODO: add support for handling inner objects easier (similar to validateRequiredField/validateOptionalField)
+    for (const priceData of value) {
+      if (typeof priceData !== 'object') {
+        throw new ValidationError(
+          'OracleSet: PriceDataSeries must be an array of objects',
+        )
+      }
+
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- we are validating the type
+      if (priceData.PriceData == null) {
```

Comment:
Should there also be a check that there's only a `PriceData` object and nothing else?

Reply:
Yes, that's a good check. Done in https://github.com/XRPLF/xrpl.js/pull/2688/commits/77481f7db74058101c78e5f94790db5fd47c37f7